### PR TITLE
[Fix] Fix nan loss in RotatedIoULoss when using AmpOptimizer

### LINKED ISSUE #888: [Bug] RTMDet/RotatedIoULoss with AmpOptimizer results in nan loss
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmrotate/issues) and [Discussions](https://github.com/open-mmlab/mmrotate/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmrotate.readthedocs.io/en/1.x/notes/4_faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmrotate) or [latest version (1.x)](https://github.com/open-mmlab/mmrotate/tree/dev-1.x).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

1.x branch https://github.com/open-mmlab/mmrotate/tree/1.x

### Environment

sys.platform: linux
Python: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:01:55) [GCC 11.3.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0,1,2,3: NVIDIA A100-PCIE-40GB
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.4, V11.4.152
GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 1.13.0.post200
PyTorch compiling details: PyTorch built with:
  - GCC 10.4
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2022.2-Product Build 20220804 for Intel(R) 64 architecture applications
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.4
  - Built with CUDA Runtime 11.2
  - NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_86,code=compute_86
  - CuDNN 8.8  (built against CUDA 11.8)
    - Built with CuDNN 8.4.1
  - Magma 2.6.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.2, CUDNN_VERSION=8.4.1, CXX_COMPILER=/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670037904586/_build_env/bin/x86_64-conda-linux-gnu-c++, CXX_FLAGS=-std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670037904586/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670037904586/work=/usr/local/src/conda/pytorch-1.13.0 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670037904586/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh=/usr/local/src/conda-prefix -isystem /usr/local/cuda/include -Wno-deprecated-declarations -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new-Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=1, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF,

TorchVision: 0.14.1a0+59d9189
OpenCV: 4.7.0
MMEngine: 0.7.4
MMRotate: 1.0.0rc1+

### Reproduces the problem - code sample

N/A

### Reproduces the problem - command or script

I use my own dataset, the config following rotated_rtmdet with mosiac and mixup augmentation
```
python tools/train.py configs/rotated_rtmdet/rotated_rtmdet_l-100e-aug-dota.py --amp
```

### Reproduces the problem - error message

```
06/26 10:15:34 - mmengine - INFO - Epoch(train)   [1][  50/1000]  lr: 6.1323e-06  eta: 1 day, 6:36:32  time: 1.1025  data_time: 0.3482  memory: 30179  loss: nan  loss_cls: 0.7615  loss_bbox: nan
06/26 10:16:13 - mmengine - INFO - Epoch(train)   [1][ 100/1000]  lr: 1.2389e-05  eta: 1 day, 1:52:00  time: 0.7618  data_time: 0.0609  memory: 30063  loss: nan  loss_cls: 0.7454  loss_bbox: nan
06/26 10:17:03 - mmengine - INFO - Epoch(train)   [1][ 150/1000]  lr: 1.8645e-05  eta: 1 day, 2:38:39  time: 1.0176  data_time: 0.2697  memory: 29996  loss: nan  loss_cls: 0.7243  loss_bbox: nan
06/26 10:17:47 - mmengine - INFO - Epoch(train)   [1][ 200/1000]  lr: 2.4901e-05  eta: 1 day, 1:58:27  time: 0.8659  data_time: 0.2335  memory: 29942  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:18:44 - mmengine - INFO - Epoch(train)   [1][ 250/1000]  lr: 3.1157e-05  eta: 1 day, 3:03:35  time: 1.1352  data_time: 0.5438  memory: 29988  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:19:31 - mmengine - INFO - Epoch(train)   [1][ 300/1000]  lr: 3.7413e-05  eta: 1 day, 2:54:43  time: 0.9475  data_time: 0.3590  memory: 30041  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:20:20 - mmengine - INFO - Epoch(train)   [1][ 350/1000]  lr: 4.3669e-05  eta: 1 day, 2:58:01  time: 0.9891  data_time: 0.4015  memory: 29941  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:21:05 - mmengine - INFO - Epoch(train)   [1][ 400/1000]  lr: 4.9926e-05  eta: 1 day, 2:42:08  time: 0.9016  data_time: 0.3191  memory: 29983  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:21:52 - mmengine - INFO - Epoch(train)   [1][ 450/1000]  lr: 5.6182e-05  eta: 1 day, 2:36:06  time: 0.9368  data_time: 0.3413  memory: 30040  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:22:38 - mmengine - INFO - Epoch(train)   [1][ 500/1000]  lr: 6.2438e-05  eta: 1 day, 2:27:47  time: 0.9167  data_time: 0.3383  memory: 29950  loss: nan  loss_cls: nan  loss_bbox: nan
06/26 10:23:12 - mmengine - INFO - Epoch(train)   [1][ 550/1000]  lr: 6.8694e-05  eta: 1 day, 1:44:48  time: 0.6774  data_time: 0.1178  memory: 30012  loss: nan  loss_cls: nan  loss_bbox: nan
```

### Additional information

When using the same dataset and the same pipeline with another head e.g. cascade RCNN with amp, the loss is not nan
When using the same dataset with hbox and the same pipeline with RTMDet in mmdet3.0 the loss is not nan
Thus, I think the problem lies in the RotatedIoULoss.
I print the RotatedIoULoss and found that some of them are not nan in a batch

### PR DESCRIPTION
**Pull request by jamiechoi1995**
Fix #888

After this fix, the loss is normal
```
06/26 10:28:07 - mmengine - INFO - Epoch(train)   [1][  50/1000]  lr: 6.1323e-06  eta: 1 day, 2:56:08  time: 0.9702  data_time: 0.1629  memory: 30334  loss: 2.0993  loss_cls: 0.8395  loss_bbox: 1.2598
06/26 10:28:45 - mmengine - INFO - Epoch(train)   [1][ 100/1000]  lr: 1.2389e-05  eta: 1 day, 0:03:42  time: 0.7640  data_time: 0.0112  memory: 29992  loss: 1.9670  loss_cls: 0.7536  loss_bbox: 1.2134
06/26 10:29:24 - mmengine - INFO - Epoch(train)   [1][ 150/1000]  lr: 1.8645e-05  eta: 23:06:32  time: 0.7654  data_time: 0.0124  memory: 30087  loss: 1.9570  loss_cls: 0.7950  loss_bbox: 1.1620
06/26 10:30:05 - mmengine - INFO - Epoch(train)   [1][ 200/1000]  lr: 2.4901e-05  eta: 23:07:00  time: 0.8360  data_time: 0.0772  memory: 30005  loss: 1.8406  loss_cls: 0.7878  loss_bbox: 1.0528
06/26 10:30:50 - mmengine - INFO - Epoch(train)   [1][ 250/1000]  lr: 3.1157e-05  eta: 23:25:05  time: 0.8904  data_time: 0.0912  memory: 30039  loss: 1.8048  loss_cls: 0.8258  loss_bbox: 0.9790
06/26 10:31:28 - mmengine - INFO - Epoch(train)   [1][ 300/1000]  lr: 3.7413e-05  eta: 23:00:47  time: 0.7599  data_time: 0.0084  memory: 30106  loss: 1.7235  loss_cls: 0.7824  loss_bbox: 0.9411
06/26 10:32:07 - mmengine - INFO - Epoch(train)   [1][ 350/1000]  lr: 4.3669e-05  eta: 22:46:50  time: 0.7751  data_time: 0.0066  memory: 30022  loss: 1.6806  loss_cls: 0.7826  loss_bbox: 0.8980
```

### CODE REVIEW

**Comment by liuyanyi** [comment]
Thanks for pr, is there any difference in mAP and training speed after using fp16?

**Comment by jamiechoi1995** [comment]
> Thanks for pr, is there any difference in mAP and training speed after using fp16?

Hi Yanyi, the mAP and training speed are nearly the same, but enabling Amp training can set a larger batch size which can reduce the training time.

**Review by RangiLyu — APPROVED** [review]
